File: Tests/Scripts/ProjectBuilder.rb
require_relative "ADB.rb"
require_relative "Tool.rb"

class ProjectBuilder < Tool

   attr_reader :binary

   def self.usage()
      tool = Tool.new()

      tool.print("\n1. Build project:", 32)
      tool.print("   $ make build\n", 36)

      tool.print("2. Enable USB Debugging on Android device. Install Android Tools for macOS. Connect Android device and Verify ADB shell setup.", 32)
      help = <<EOM
   $ make verify

   References:
   - How to Install Android Tools for macOS: https://stackoverflow.com/questions/17901692/set-up-adb-on-mac-os-x
   - How to Enable USB Debugging on Android device: https://developer.android.com/studio/debug/dev-options
EOM
      tool.print(help, 36)

      tool.print("3. Deploy and run project on Android Device or Simulator.", 32)
      help = <<EOM
   $ make deploy:armv7a
   $ make deploy:aarch64
   $ make deploy:x86
   $ make deploy:x86_64
EOM
      tool.print(help, 36)

      tool.print("4. (Optional) Clean deployed project:", 32)
      tool.print("   $ make clean:armv7a", 36)
      tool.print("   $ make clean:aarch64", 36)
      tool.print("   $ make clean:x86", 36)
      tool.print("   $ make clean:x86_64", 36)

      tool.print("\n5. (Optional) Clean project:", 32)
      tool.print("   $ make clean\n", 36)
   end

   def self.perform()
      action = ARGV.first
      if action.nil? then usage()
      elsif action == "build" then build()
      elsif action == "clean" then clean()
      elsif action == "verify" then ADB.verify()
      elsif action.start_with?("clean:") then undeploy(action.sub("clean:", ''))
      elsif action.start_with?("deploy:") then deploy(action.sub("deploy:", ''))
      else usage()
      end
   end

   def self.build()
      Tool.new().archsToBuild.each { |arch| Builder.new(arch).build() }
   end

   def self.clean()
      Tool.new().archsToBuild.each { |arch| Builder.new(arch).clean() }
   end

   def self.deploy(arch)
      Builder.new(arch).deploy()
   end

   def self.undeploy(arch)
      Builder.new(arch).undeploy()
   end

   def initialize(component, arch)
      @isVerbose = false
      @arch = arch
      @component = component
      @sources = "#{@root}/Sources"
      @builds = "#{@root}/build-#{arch}"
      @config = "release"
      @toolchainDir = File.expand_path(File.join(File.dirname(__FILE__), "../../ToolChain/swift-android-toolchain"))
      if @arch == "armv7a"
         @ndkArchPath = "arm-linux-androideabi"
         @swiftTarget = "armv7-none-linux-androideabi"
      elsif @arch == "x86"
         @ndkArchPath = "i686-linux-android"
         @swiftTarget = "i686-unknown-linux-android"
      elsif @arch == "aarch64"
         @ndkArchPath = "aarch64-linux-android"
         @swiftTarget = "aarch64-unknown-linux-android"
      elsif @arch == "x86_64"
         @ndkArchPath = "x86_64-linux-android"
         @swiftTarget = "x86_64-unknown-linux-android"
      end
      @swiftc = @toolchainDir + "/usr/bin/swiftc-" + @ndkArchPath
      @copyLibsCmd = @toolchainDir + "/usr/bin/copy-libs-" + @ndkArchPath
      @swiftBuildCmd = "#{@toolchainDir}/usr/bin/android-swift-build --android-target #{@swiftTarget} -c #{@config} --build-path #{@builds}"

      if @isVerbose
         @swiftc += " -v"
         @copyLibsCmd += " -v"
         @swiftBuildCmd += " -v"
      end
   end

   def libs
      return Dir["#{@builds}/lib/*"]
   end

   def copyLibs()
      targetDir = "#{@builds}/lib"
      execute "rm -rf \"#{targetDir}\""
      execute "#{@copyLibsCmd} #{targetDir}"
   end

   def build
      clean()
      execute "mkdir -p \"#{@builds}\""
      executeBuild()
   end

   def clean
      execute "rm -rf \"#{@builds}\""
   end

   def deploy()
      adb = ADB.new(libs, binary)
      adb.deploy()
      adb.run()
   end

   def undeploy()
      ADB.new(libs, binary).clean
   end

end

File: Tests/ProjectBuilder.js
var path = require("path");
const cp = require("child_process");

const Tool = require("../lib/Tool");
const Arch = require("../lib/Archs/Arch");
const Config = require("../lib/Config");
const ADB = require("./ADB");

module.exports = class ProjectBuilder extends Tool {
  constructor(/** @type {String} */ component, /** @type {Arch} */ arch) {
    super();
    var args = process.argv.slice(2);
    this.isVerbose = args.filter((item) => item == "--verbose").length > 0;

    this.arch = arch;
    this.component = component;
    this.rootPath = path.join(Config.root, "Tests", `sample-${component}`);
    this.sourcesPath = path.join(this.rootPath, "Sources");
    this.buildPath = path.join(this.rootPath, "build", arch.name);
    this.binariesDirPath = path.join(Config.root, "Assets");
    if (arch.name == "armv7a") {
      this.ndkArchPath = "arm-linux-androideabi";
      this.swiftTarget = "armv7-none-linux-androideabi";
    } else if (arch.name == "x86") {
      this.ndkArchPath = "i686-linux-android";
      this.swiftTarget = "i686-unknown-linux-android";
    } else if (arch.name == "aarch64") {
      this.ndkArchPath = "aarch64-linux-android";
      this.swiftTarget = "aarch64-unknown-linux-android";
    } else if (arch.name == "x86_64") {
      this.ndkArchPath = "x86_64-linux-android";
      this.swiftTarget = "x86_64-unknown-linux-android";
    }
    this.swftcCmdPath = `${this.binariesDirPath}/swiftc-${this.ndkArchPath} -module-cache-path ${this.buildPath}/ModuleCache`;
    this.copyLibsCmdPath = `${this.binariesDirPath}/copy-libs-${this.ndkArchPath}`;

    this.swiftBuildCmd = `${this.binariesDirPath}/swift-build-${this.ndkArchPath} --build-path "${this.buildPath}"`;

    if (this.isVerbose) {
      this.swftcCmdPath += " -v -Xcc -v";
      this.copyLibsCmdPath += " -v";
      this.swiftBuildCmd += " -v";
    }

    this.binary = `${this.buildPath}/${this.component}`;
    this.libsDirPath = path.join(this.buildPath, "libs");
  }

  runAction(/** @type {String} */ action) {
    if (action == "build") {
      this.build();
    } else if (action == "deploy") {
      this.deploy();
    } else if (action == "undeploy") {
      this.undeploy();
    } else if (action == "clean") {
      this.clean();
    } else {
      this.logError(`! Unknown action \"${action}\".`);
      process.exit(1);
    }
  }

  executeBuild() {}

  copyLibs() {
    this.execute(`rm -rf "${this.libsDirPath}"`);
    this.execute(`${this.copyLibsCmdPath} -output "${this.libsDirPath}"`);
  }

  get libs() {
    var libs = cp.execSync(`find "${this.libsDirPath}" -iname *.so -depth 1 || true`).toString().trim().split("\n");
    return libs;
  }

  deploy() {
    let adb = new ADB(this.binary, this.libs);
    adb.deploy();
    adb.run();
  }

  undeploy() {
    new ADB(this.binary, this.libs).undeploy();
  }

  build() {
    this.clean();
    this.execute(`mkdir -p "${this.buildPath}"`);
    this.executeBuild();
  }

  clean() {
    this.execute(`rm -rf "${this.buildPath}"`);
  }
};
